sklearn/model_selection/_classification_threshold.py

```diff
@@ -996,8 +1008,15 @@ def get_metadata_routing(self):
 
     def _get_curve_scorer(self):
         """Get the curve scorer based on the objective metric used."""
-        scoring = check_scoring(self.estimator, scoring=self.scoring)
+        scorer = check_scoring(self.estimator, scoring=self.scoring)
+        if scorer._response_method != "predict":
+            raise ValueError(
```

Comment:
Here, I’m not so sure whether we overly constrain possible scorers from the user. But I also do not 100% follow how the curve scorer works.

Reply:
If we don't do that and the users pass a non thresholded metric like roc auc/average precision/log loss/brier, the metric is evaluated on the thresholded predictions which is really misleading. You still get a 'tuned' threshold but its meaning is really confusing.

Reply:
I think that we can first restrain and be conservative. If a real use case is reported then we can then rework and make sure that the API is right and not just working as a side effect.

Reply:
I don't like that here we're accessing private `_response_method` as a proxy to see if it's the right _kind_ of scorer or not. Scorers should have a public API for this (cc @StefanieSenger )

Reply:
It seems to me that since `response_method="predict"` is the default in `_BaseScorer`, this error might not reach the users that might be most vulnerable to be confused by wrong outputs. Is there a more computed way to check if the scorer fits the purpose that does not depend on user input?

Reply:
Let me know if the following helps address those concerns:

https://github.com/scikit-learn/scikit-learn/pull/29082/files#r1613325997